### PR TITLE
EWPP-141: Test coverage.

### PR DESCRIPTION
**Pull request by julien-**

### CODE REVIEW

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    $this->drupalLogin($this->blockVisibilityGroupsUser);
+    // Enable action and task blocks.
+    $this->drupalPlaceBlock('local_actions_block');
+    $this->drupalPlaceBlock('local_tasks_block');
+
+    // Tab exists and works.
+    $this->drupalGet('admin/structure/block');
+    $this->assertLink('Block Visibility Groups');
+    $this->clickLink('Block Visibility Groups');
+    $this->assertResponse(200);
```
Comment (yenyasinn) [inline tests/src/Functional/BlockVisibilityGroupsUITest.php:27]: I would suggest to ensure that we are on the correct page:
```
    $this->assertSession()->statusCodeEquals(200);
    $this->assertSession()->pageTextContains('Block layout');
```

Comment (yenyasinn) [inline tests/src/Functional/BlockVisibilityGroupsUITest.php:27]: assertResponse is deprecated. Use $this->assertSession()->statusCodeEquals() instead.

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  * Test for the edition of block visibility groups.
+  */
+  public function testBlockVisibilityEditForm() {
+    $this->drupalLogin($this->blockVisibilityGroupsUser);
+    $id = 'test_block_visibility_groups';
+    $label = $this->randomMachineName();
+    $this->createBlockVisibilityGroup($id, $label);
+    // Edit test with new label.
+    $this->assertSession()->elementTextContains('css', 'main table .dropbutton', 'Edit');
+    $this->drupalGet('admin/structure/block/block-visibility-group/'.$id);
```
Comment (yenyasinn) [inline tests/src/Functional/BlockVisibilityGroupsUITest.php:54]: Drupal coding standard issue. Have to be
```
$this->drupalGet('admin/structure/block/block-visibility-group/ ' . $id);
```

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  * Test for the deletion of block visibility groups.
+  */
+  public function testBlockVisibilityDelete() {
+    $this->drupalLogin($this->blockVisibilityGroupsUser);
+    $id = 'test_block_visibility_groups';
+    $label = $this->randomMachineName();
+    $this->createBlockVisibilityGroup($id, $label);
+    // Delete test.
+    $this->assertSession()->elementTextContains('css', 'main table .dropbutton', 'Delete');
+    $this->drupalGet('admin/structure/block/block-visibility-group/'.$id.'/delete');
```
Comment (yenyasinn) [inline tests/src/Functional/BlockVisibilityGroupsUITest.php:71]: Please add spaces around dots.

File: tests/src/Functional/BasicTest.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+   * Create a new block visibility group through the UI.
+   *
+   * @param string $_id
+   *   The machine name of the group.
+   * @param string $label
+   *   The label of the group.
+   */
+  protected function createBlockVisibilityGroup($id, $label){
```
Comment (drishu) [inline tests/src/Functional/BasicTest.php:87]:
```suggestion
  protected function createBlockVisibilityGroup($id, $label) {
```

File: tests/src/Functional/BlockVisibilityGroupsTestBase.php [hunk 5]
```diff
@@ -105,4 +114,20 @@ protected function updateBlockInGroupUI($block_id, $group_id, array $settings =
...
 
+  /**
+   * Create a new block visibility group through the UI.
+   *
+   * @param string $_id
+   *   The machine name of the group.
+   * @param string $label
+   *   The label of the group.
+   */
+  protected function createBlockVisibilityGroup($id, $label){
```
Comment (drishu) [inline tests/src/Functional/BlockVisibilityGroupsTestBase.php:125]:
```suggestion
  protected function createBlockVisibilityGroup($id, $label) {
```

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Tests the block_visibility_groups UI.
+ *
+ * @group block_visibility_groups
+ */
+class BlockVisibilityGroupsUITest extends BlockVisibilityGroupsTestBase {
+
+  /**
+   * Test for the creation of block visibility groups.
```
Comment (drishu) [inline tests/src/Functional/BlockVisibilityGroupsUITest.php:15]: this is not what we are testing here

File: tests/src/Functional/VisibilityTest.php [hunk 7]
```diff
@@ -40,10 +36,9 @@ protected function setUp() {
   public static $modules = ['block', 'block_visibility_groups', 'node'];
 
   /**
-   * {@inheritdoc}
+   * Test single conditions.
    */
   public function testSingleConditions() {
-    // @todo Condition with node doesn't work for some reason.
     $config = [
```
Comment (drishu) [inline tests/src/Functional/VisibilityTest.php:42]: is this bit used?

File: tests/src/FunctionalJavascript/BlockVisibilityGroupsTestBase.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    parent::setUp();
+    
+    // User can administer blocks from core.
+    $this->blockVisibilityGroupsUser = $this->drupalCreateUser([
+      'administer blocks',
+      'administer block visibility groups',
+    ]);
+  }
+
+}
```
Comment (drishu) [inline tests/src/FunctionalJavascript/BlockVisibilityGroupsTestBase.php:45]: There is a lot missing here, looking at the module I would say we need to test:
- block visibility groups crud 
- conditions crud: we need to test ..condition/add/ /edit /delete
- assigning bvg to block
- test admin page filter
- test that conditions apply frontend (block visibility is taken into account for different settings)

File: tests/src/Functional/BlockVisibilityGroupsUITest.php (hunk 6 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:15]: FIxed. Yes it was a duplication (I keep it as backup ... sorry)

File: tests/src/Functional/BasicTest.php (hunk 4 above)
Comment (julien-) [reply tests/src/Functional/BasicTest.php:87]: Fixed. YES it was. I remove BasicTest.php

File: tests/src/Functional/BlockVisibilityGroupsUITest.php (hunk 1 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:27]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    $this->clickLink('Add Block Visibility Group');
+    $this->assertResponse(200);
+    $this->assertFieldById('edit-label');
+    $label = $this->randomMachineName();
+    $this->drupalPostForm(NULL, [
+      'label' => $label,
+      'id' => 'test_block_visibility_groups',
+    ], 'Save');
+    $this->assertResponse(200);
+    $this->assertText(new FormattableMarkup('Saved the @group Block Visibility Group.', ['@group' => $label]));
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:40]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsUITest.php (hunk 1 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:27]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsUITest.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    $this->drupalPlaceBlock('local_actions_block');
+    $this->drupalPlaceBlock('local_tasks_block');
+
+    // Tab exists and works.
+    $this->drupalGet('admin/structure/block');
+    $this->assertLink('Block Visibility Groups');
+    $this->clickLink('Block Visibility Groups');
+    $this->assertResponse(200);
+
+    // Add test.
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:29]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsUITest.php (hunk 2 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:54]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsUITest.php (hunk 3 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsUITest.php:71]: Fixed

File: tests/src/Functional/BlockVisibilityGroupsTestBase.php (hunk 5 above)
Comment (julien-) [reply tests/src/Functional/BlockVisibilityGroupsTestBase.php:125]: Fixed

**Pull request merged**